grpc/client.go

```diff
@@ -0,0 +1,363 @@
+				c.RequestTimeout, minRequestTimeout)
+		}
+	}
+
+	return nil
+}
+
+// RetryConfig denotes the gRPC retry policy
+type RetryConfig struct {
+	// InitialBackoff is the initial delay before retrying a request
+	InitialBackoff types.Duration `mapstructure:"InitialBackoff"`
+
+	// MaxBackoff is the maximum backoff duration for retries
+	MaxBackoff types.Duration `mapstructure:"MaxBackoff"`
+
+	// BackoffMultiplier is the multiplier for the backoff duration
+	BackoffMultiplier float64 `mapstructure:"BackoffMultiplier"`
+
+	// MaxAttempts is the maximum number of retries for a request
+	MaxAttempts int `mapstructure:"MaxAttempts"`
+
+	// Excluded captures functions which are excluded from retry policies
+	Excluded []Method `mapstructure:"Excluded"`
+}
+
+type Method struct {
+	// ServiceName identifies gRPC service name (alongside package)
+	ServiceName string `mapstructure:"Service"`
+
+	// MethodName denotes gRPC function name
+	MethodName string `mapstructure:"Method"` // optional
+}
+
+func (r *RetryConfig) String() string {
+	if r == nil {
+		return noneStr
+	}
+
+	return fmt.Sprintf("InitialBackoff=%s, MaxBackoff=%s, "+
+		"BackoffMultiplier=%f, MaxAttempts=%d",
+		r.InitialBackoff.String(), r.MaxBackoff.String(),
+		r.BackoffMultiplier, r.MaxAttempts,
+	)
+}
+
+// Validate checks if the gRPC retry configuration is valid.
+// It returns an error if any of the required fields are missing or invalid.
+func (r *RetryConfig) Validate() error {
+	if r.InitialBackoff.Duration <= 0 {
+		return fmt.Errorf("InitialBackoff must be greater than zero")
+	}
+
+	if r.MaxBackoff.Duration <= 0 {
+		return fmt.Errorf("MaxBackoff must be greater than zero")
+	}
+
+	if r.InitialBackoff.Duration >= r.MaxBackoff.Duration {
+		return fmt.Errorf("InitialBackoff must be less than MaxBackoff")
+	}
+
+	if r.BackoffMultiplier < 1.0 {
+		return fmt.Errorf("BackoffMultiplier must be greater than 1.0")
+	}
+
+	if r.MaxAttempts < 1 {
+		return fmt.Errorf("MaxAttempts must be at least 1")
+	}
+
+	return nil
+}
+
+// Client holds the gRPC connection and services
+type Client struct {
+	conn *grpc.ClientConn
+}
+
+// NewClient initializes and returns a new gRPC client
+func NewClient(cfg *ClientConfig) (*Client, error) {
+	if cfg == nil {
+		return nil, fmt.Errorf("gRPC client configuration cannot be nil")
+	}
+
+	dialBackoff := backoff.DefaultConfig
+	retryCfg := cfg.Retry
+	if retryCfg != nil {
+		dialBackoff.BaseDelay = retryCfg.InitialBackoff.Duration
+		dialBackoff.MaxDelay = retryCfg.MaxBackoff.Duration
+		dialBackoff.Multiplier = retryCfg.BackoffMultiplier
+	}
+	connectParams := grpc.ConnectParams{
+		Backoff:           dialBackoff,
+		MinConnectTimeout: cfg.MinConnectTimeout.Duration,
+	}
+
+	serviceCfgJSON, err := createServiceConfig(retryCfg)
+	if err != nil {
+		return nil, err
+	}
+
+	opts := []grpc.DialOption{
+		grpc.WithTransportCredentials(insecure.NewCredentials()),
+		grpc.WithConnectParams(connectParams),
+		grpc.WithDefaultServiceConfig(serviceCfgJSON),
```

Comment:
I am not sure what are the consequences if you put default service config as empty string. Maybe just add it if it is not empty, just to be safe.